### PR TITLE
Built in profiling

### PR DESCRIPTION
**Pull request by BaronKiko**
This PR attempts to implement a profiler within ryujinx.  Currently it only profiles system services and I want to know what you want profiled. It's only a couple lines to add one so don't be shy.
Honestly it's not the nicest UI but it's pretty lightweight and not really intended for end users but devs to be able to find potential bottlenecks.

It looks like this (slightly out of date but you get the idea):
ryujinx_2019-01-26_21-14-27

Features (so you know what to test):

- F1 (change in config) to hide/show the profiler. Only works when main window focused
- Click column heading to sort
- Scrollable data
- Hide inactive button
- Play/Pause button
- Step button when paused
- Filtering, type to change filter with profiler focused
- Regex filtering, click filter bar at bottom to toggle
- Entries should be searchable as Category.Group.Item. For example you can get only the top entry in the picture above with "ServiceCall.ISharedFontManager.GetFontSize" or emit GetFontSize to get all children of ISharedFontManager
- Timing bars render correctly (you have to switch to it with view button)
- No idea if mac or linux work (can someone who has them set up test?)
- Changing update rate from config
- CSV dumping, only enabled with a valid path specified in config.
- Graphing, you should be able to scroll/zoom with arrow
- Flags, the vertical lines that appear in graph mode. They represent a frame swap atm but you can add more later. Should probably draw them in different colours but there is only 1 at the time of writing this.
- Maybe some other things, hard to remember all I did

Known issues:

- Need a better way to get default font locations

I think that's all I have to say atm. Hopefully it works for you.

### CODE REVIEW

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 1]
```diff
@@ -0,0 +1,545 @@
...
+
+            FontServ.fontColor = Color.White;
+            int verticalIndex = 0;
+            int lineHeight = 16;
+            int titleHeight = 24;
+            int titleFontHeight = 16;
+            int linePadding = 2;
+            int columnSpacing = 30;
+            int filterHeight = 24;
+
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align, below too

Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align, below too

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 2]
```diff
@@ -0,0 +1,545 @@
...
+            // Background lines to make reading easier
+            GL.Enable(EnableCap.ScissorTest);
+            GL.Scissor(0, filterHeight, Width, Height - titleHeight - filterHeight);
+            GL.Begin(PrimitiveType.Triangles);
+            GL.Color3(0.2f, 0.2f, 0.2f);
+            for (int i = 0; i < ProfileData.Count; i += 2)
+            {
+                float top = GetLineY(yOffset, lineHeight, linePadding, false, i - 1);
+                float bottom = GetLineY(yOffset, lineHeight, linePadding, false, i);
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 3]
```diff
@@ -0,0 +1,545 @@
...
+                }
+            }
+            GL.Disable(EnableCap.ScissorTest);
+
+            width = FontServ.DrawText("Category", xOffset, Height - titleFontHeight, titleFontHeight);
+            if (width > maxWidth)
+                maxWidth = width;
+
+            xOffset += maxWidth + columnSpacing;
+
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: remove whitespace

Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: remove whitespace

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 4]
```diff
@@ -0,0 +1,545 @@
...
+            width = FontServ.DrawText("Category", xOffset, Height - titleFontHeight, titleFontHeight);
+            if (width > maxWidth)
+                maxWidth = width;
+
+            xOffset += maxWidth + columnSpacing;
+            
+
+            // Display session group
+            maxWidth = 0;
+            verticalIndex = 0;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 5]
```diff
@@ -0,0 +1,545 @@
...
+
+            width = FontServ.DrawText("Group", xOffset, Height - titleFontHeight, titleFontHeight);
+            if (width > maxWidth)
+                maxWidth = width;
+
+            xOffset += maxWidth + columnSpacing;
+
+            // Display session item
+            maxWidth = 0;
+            verticalIndex = 0;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 6]
```diff
@@ -0,0 +1,545 @@
...
+                maxWidth = width;
+
+            xOffset += maxWidth + columnSpacing;
+            Buttons[(int)ButtonIndex.TagTitle].UpdateSize(0, Height - titleFontHeight, 0, (int)xOffset, titleFontHeight);
+
+            // Time bars
+            if (ProfileData.Count != 0)
+            {
+                width = Width - xOffset - 370;
+                long maxAverage, maxTotal;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: declare these at the beginning of this scope,  each variable on a new line

Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: declare these at the beginning of this scope,  each variable on a new line

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 7]
```diff
@@ -0,0 +1,545 @@
...
+                float barHeight = (lineHeight - linePadding) / 3.0f;
+                verticalIndex = 0;
+
+                // Get max values
+                var maxInstant = maxAverage = maxTotal = 0;
+                foreach (KeyValuePair<ProfileConfig, TimingInfo> kvp in ProfileData)
+                {
+                    maxInstant = Math.Max(maxInstant, kvp.Value.Instant);
+                    maxAverage = Math.Max(maxAverage, kvp.Value.AverageTime);
+                    maxTotal = Math.Max(maxTotal, kvp.Value.TotalTime);
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindow.cs (hunk 6 above)
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: declare these at the beginning of this scope,  each variable on a new line

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 8]
```diff
@@ -0,0 +1,545 @@
...
+        protected override void OnMouseUp(MouseButtonEventArgs e)
+        {
+            foreach (ProfileButton button in Buttons)
+            {
+                if (button.ProcessClick(e.X, Height - e.Y))
+                    return;
+            }
+        }
+
+        protected override void OnMouseMove(MouseMoveEventArgs e)
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: if not abstract, remove

File: Ryujinx.Profiler/UI/ProfileWindowManager.cs [hunk 9]
```diff
@@ -0,0 +1,44 @@
+﻿using System.Threading;
+
+namespace Ryujinx.Profiler.UI
+{
+    public class ProfileWindowManager
+    {
+        private ProfileWindow Window;
+        private Thread ProfileThread;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindowManager.cs]: _camelCase for private fields

File: Ryujinx.Profiler/UI/ProfileWindow.cs (hunk 8 above)
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: if not abstract, remove

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+namespace Ryujinx.Profiler.UI.SharpFontHelpers
+{
+    public class FontService
+    {
+        private struct CharacterInfo
+        {
+            public float left, right, bottom, top;
+            public float height, aspectRatio, xBearing, yBearing, advance;
+            public int width;
+        }
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: each variable on new line

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+        private int CharacterTextureSheet;
+        private CharacterInfo[] characters;
+
+        public Color fontColor = Color.Black;
+
+        public void InitalizeTextures()
+        {
+            // Create and init some vars
+            uint[] rawCharacterSheet = new uint[SheetWidth * SheetHeight];
+            int x, y, lineOffset, maxHeight;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: variable per line

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+            characters = new CharacterInfo[94];
+
+            // Get font
+            var font = new FontFace(File.OpenRead(Path.Combine(Environment.GetFolderPath(
+                Environment.SpecialFolder.ApplicationData), @"RyuFs\system\fonts\FontStandard.ttf")));
+
+            // Update raw data for each character
+            for (int i = 0; i < 94; i++)
+            {
+                float xBearing, yBearing, advance;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: this

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 13]
```diff
@@ -0,0 +1,230 @@
...
+            }
+
+            // Fix height for characters shorter than line height
+            for (int i = 0; i < 94; i++)
+            {
+                characters[i].xBearing /= characters[i].width;
+                characters[i].yBearing /= maxHeight;
+                characters[i].advance /= characters[i].width;
+                characters[i].height /= maxHeight;
+                characters[i].aspectRatio = (float)characters[i].width / maxHeight;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 14]
```diff
@@ -0,0 +1,230 @@
...
+            }
+
+            // Convert raw data into texture
+            CharacterTextureSheet = GL.GenTexture();
+            GL.BindTexture(TextureTarget.Texture2D, CharacterTextureSheet);
+
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMinFilter, (int)TextureMinFilter.Linear);
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureMagFilter, (int)TextureMagFilter.Linear);
+            GL.TexParameter(TextureTarget.Texture2D, TextureParameterName.TextureWrapS, (int)TextureWrapMode.Clamp);
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align end

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 15]
```diff
@@ -0,0 +1,230 @@
...
+            GL.TexCoord2(charInfo.right, charInfo.bottom); GL.Vertex2(right, bottom);
+            GL.TexCoord2(charInfo.left, charInfo.bottom);  GL.Vertex2(left, bottom);
+        }
+
+        public unsafe Surface RenderSurface(char c, FontFace font, out float xBearing, out float yBearing, out float advance)
+        {
+            var glyph = font.GetGlyph(c, 32);
+            xBearing = glyph.HorizontalMetrics.Bearing.X;
+            yBearing = glyph.RenderHeight - glyph.HorizontalMetrics.Bearing.Y;
+            advance = glyph.HorizontalMetrics.Advance;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+
+            var surface = new Surface
+            {
+                Bits = Marshal.AllocHGlobal(glyph.RenderWidth * glyph.RenderHeight),
+                Width = glyph.RenderWidth,
+                Height = glyph.RenderHeight,
+                Pitch = glyph.RenderWidth
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+
+            return surface;
+        }
+
+        private CharacterInfo UpdateTexture(Surface surface, ref uint[] rawCharMap, ref int posX, ref int posY, ref int lineOffset)
+        {
+            int width = surface.Width;
+            int height = surface.Height;
+            int len = width * height;
+            byte[] data = new byte[len];
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 18]
```diff
@@ -0,0 +1,230 @@
...
+            if (lineOffset < height)
+            {
+                lineOffset = height + 1;
+            }
+
+            // Copy char to sheet
+            for (int y = 0; y < height; y++)
+            {
+                int destOffset = (y + posY) * SheetWidth + posX;
+                int sourceOffset = y * width;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 19]
```diff
@@ -0,0 +1,230 @@
...
+            // Generate character info
+            CharacterInfo charInfo = new CharacterInfo()
+            {
+                left = (float)posX / SheetWidth,
+                right = (float)(posX + width) / SheetWidth,
+                top = (float)(posY - 1) / SheetHeight,
+                bottom = (float)(posY + height) / SheetHeight,
+                width = width,
+                height = height,
+            };
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: align

**Comment by BaronKiko** [comment]
@emmauss Waiting for that :) There is enough code that it will probably take a couple passes. Anyway I will review your suggestions after dinner, and maybe sleep.

File: Ryujinx.Profiler/InternalProfile.cs [hunk 20]
```diff
@@ -0,0 +1,105 @@
...
+
+namespace Ryujinx.Profiler
+{
+    public class InternalProfile
+    {
+        private Stopwatch SW;
+        internal ConcurrentDictionary<ProfileConfig, TimingInfo> Timers;
+
+        private readonly object SessionLock = new object();
+        private int SessionCounter = 0;
```
[the remark this answers is not in the document]
Comment (BaronKiko) [reply Ryujinx.Profiler/InternalProfile.cs]: Internal should be Pascal according to Gdk, otherwise done. Commit coming later,

File: Ryujinx.Profiler/Profile.cs [hunk 21]
```diff
@@ -0,0 +1,86 @@
...
+        }
+
+        public static void Method(ProfileConfig config, Action method)
+        {
+            // If profiling is disabled just call the method
+            if (!ProfilingEnabled())
+                method();
+
+            Begin(config);
+            method();
```
[the remark this answers is not in the document]
Comment (BaronKiko) [reply Ryujinx.Profiler/Profile.cs]: > pascal case for method. If there is a naming conflict, think of a better name. `method` is too generic

Not exactly sure what you mean here.
The idea of this function is it's called with Profie.Method(some_config, () => do something);
So you are profiling a method, it can be anything so the name is meant to be very generic.
If you are referring to the variable it's a local var so lower case is right?

File: Ryujinx.Profiler/Settings.cs [hunk 22]
```diff
@@ -0,0 +1,14 @@
...
+
+namespace Ryujinx.Profiler
+{
+    public class ProfilerSettings
+    {
+        public bool Enabled = true;
+        public bool FileDumpEnabled = false;
+        public string DumpLocation = "";
+        public float UpdateRate = 0.1f;
+    }
```
[the remark this answers is not in the document]
Comment (BaronKiko) [reply Ryujinx.Profiler/Settings.cs]: Not sure what you mean by use auto props as they don't really apply here. Those are just the defaults. I have added a comment to clarify that they are defaults and aligned them

**Comment by BaronKiko** [comment]
Stopped using window.Run so it should be easier on your cpu/gpu.
It also now only renders a frame when there is actually new data/ui interaction to display
@emmauss let me know how it goes for you, I hope it's worth it because it was a pain

**Comment by BaronKiko** [comment]
Huge update, adds real time bar graph support:
image
It also supports TimingFlags (the vertical bars), I have currently set them up for ogl frame swap so you can clearly see inconsistent frame times.
You can change update rate to 60fps in config to get proper real time but it may give you performance issues as it's processing a lot of data.
Anyway enjoy

File: Ryujinx.Profiler/Settings.cs (hunk 22 above)
[the remark this answers is not in the document]
Comment (jduncanator) [reply Ryujinx.Profiler/Settings.cs]: General rule is to make any public member a property, those are currently public fields.

**Comment by BaronKiko** [comment]
I think that's basically it for the actual profiler but I still need to know what to profile and ideally sort out the system font issue as looking at the VirtualFileSystem class I'm just more confused. If I understood it correctly no code change is needed.

**Comment by BaronKiko** [comment]
Please review for merge :)

File: Ryujinx.HLE/HOS/Services/IpcService.cs [hunk 23]
```diff
@@ -95,7 +96,13 @@ public void CallMethod(ServiceCtx context)
 
                 Logger.PrintDebug(LogClass.KernelIpc, $"{service.GetType().Name}: {processRequest.Method.Name}");
 
+                ProfileConfig profile = Profiles.ServiceCall;
+                profile.SessionGroup = service.GetType().Name;
+                profile.SessionItem = processRequest.Method.Name;
```
Comment (emmauss) [inline Ryujinx.HLE/HOS/Services/IpcService.cs]: align

File: Ryujinx.Profiler/Settings.cs [hunk 24]
```diff
@@ -0,0 +1,22 @@
...
+    {
+        // Default settings for profiler
+        public bool   Enabled         { get; set; } = false;
+        public bool   FileDumpEnabled { get; set; } = false;
+        public string DumpLocation    { get; set; } = "";
+        public float  UpdateRate      { get; set; } = 0.1f;
+        public int    MaxLevel        { get; set; } = 0;
+
+        // 19531225 = 5 seconds in ticks
+        public long   History         { get; set; } = 19531225;
```
Comment (emmauss) [inline Ryujinx.Profiler/Settings.cs]: no need for the spaces.

File: Ryujinx.Profiler/TimingInfo.cs [hunk 25]
```diff
@@ -0,0 +1,175 @@
...
+        // Timestamp collection
+        private List<Timestamp> _timestamps;
+        private readonly object _timestampLock     = new object();
+        private readonly object _timestampListLock = new object();
+        private Timestamp _currentTimestamp;
+
+        // Depth of current timer,
+        // each begin call increments and each end call decrements
+        private int _depth;
+
```
Comment (emmauss) [inline Ryujinx.Profiler/TimingInfo.cs]: extra new line

File: Ryujinx.Profiler/UI/ProfileButton.cs [hunk 26]
```diff
@@ -0,0 +1,111 @@
...
+
+        // Label information
+        private int    _labelX, _labelY;
+        private string _label;
+
+        // Misc
+        private Action _clicked;
+        private bool   _visible;
+
+
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileButton.cs]: extra new line

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 27]
```diff
@@ -0,0 +1,726 @@
...
+            Autodraw          = 8,
+        }
+
+        // Font service
+        private FontService _fontService;
+
+        // UI variables
+        private ProfileButton[] _buttons;
+
+        private bool _initComplete    = false;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: no  need for the new lines below. the can be grouped together

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 28]
```diff
@@ -0,0 +1,726 @@
...
+        private float _minScroll = 0;
+        private float _maxScroll = 0;
+
+        // Profile data storage
+        private List<KeyValuePair<ProfileConfig, TimingInfo>> _sortedProfileData;
+        private long _captureTime;
+
+        // Input
+        private bool _backspaceDown       = false;
+        private bool _prevBackspaceDown   = false;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 29]
```diff
@@ -0,0 +1,726 @@
...
+
+        private Key _graphControlKey = Key.F35;
+
+        // Event management
+        private double _updateTimer;
+        private double _processEventTimer;
+        private bool   _profileUpdated           = false;
+        private readonly object _profileDataLock = new object();
+        
+
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: no need for a new line

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 30]
```diff
@@ -0,0 +1,726 @@
...
+                _fontService.DrawText("Average", 150 + xOffset, yHeight, TitleFontHeight);
+                _buttons[(int)ButtonIndex.AverageTitle].UpdateSize((int)(150 + xOffset), (int)yHeight, 0, 130, TitleFontHeight);
+
+                _fontService.DrawText("Total (ms)", 260 + xOffset, yHeight, TitleFontHeight);
+                _buttons[(int)ButtonIndex.TotalTitle].UpdateSize((int)(260 + xOffset), (int)yHeight, 0, Width, TitleFontHeight);
+
+                // Totals
+                yHeight = FilterHeight + 2;
+                _fontService.DrawText($"{totalInstant / PerformanceCounter.TicksPerMillisecond:F3} ({totalCount})", xOffset, yHeight, TitleFontHeight);
+                _fontService.DrawText($"{totalAverage / PerformanceCounter.TicksPerMillisecond:F3}", 150 + xOffset, yHeight, TitleFontHeight);
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: align the params

File: Ryujinx.Profiler/UI/ProfileWindowBars.cs [hunk 31]
```diff
@@ -0,0 +1,83 @@
...
+                int verticalIndex = 0;
+                float barHeight   = (LineHeight - LinePadding) / 3.0f;
+
+                // Get max values
+                var maxInstant = maxAverage = maxTotal = 0;
+                foreach (KeyValuePair<ProfileConfig, TimingInfo> kvp in _sortedProfileData)
+                {
+                    maxInstant = Math.Max(maxInstant, kvp.Value.Instant);
+                    maxAverage = Math.Max(maxAverage, kvp.Value.AverageTime);
+                    maxTotal = Math.Max(maxTotal, kvp.Value.TotalTime);
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindowBars.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindowBars.cs [hunk 32]
```diff
@@ -0,0 +1,83 @@
...
+                }
+
+                GL.Enable(EnableCap.ScissorTest);
+                GL.Begin(PrimitiveType.Triangles);
+                foreach (var entry in _sortedProfileData)
+                {
+                    // Instant
+                    GL.Color3(Color.Blue);
+                    float bottom = GetLineY(yOffset, LineHeight, LinePadding, true, verticalIndex++);
+                    float top = bottom + barHeight;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindowBars.cs]: align

File: Ryujinx.Profiler/UI/ProfileWindowBars.cs [hunk 33]
```diff
@@ -0,0 +1,83 @@
...
+
+                    GL.Vertex2(right, top);
+                    GL.Vertex2(right, bottom);
+                    GL.Vertex2(xOffset, bottom);
+
+                    // Total
+                    GL.Color3(Color.Red);
+                    top += barHeight;
+                    bottom += barHeight;
+                    right = (float)entry.Value.TotalTime / maxTotal * width + xOffset;
```
Comment (emmauss) [inline Ryujinx.Profiler/UI/ProfileWindowBars.cs]: align, and maybe a new line after this line.

**Comment by BaronKiko** [comment]
Fixed formatting issues left
Added frame timing display to cover this commit closing: https://github.com/Ryujinx/Ryujinx/pull/551
Coloured flags to differentiate between flag types
Show/Hide flag button added
Needs another review

**Comment by BaronKiko** [comment]
Bump maybe?

File: Ryujinx.HLE/PerformanceStatistics.cs [hunk 34]
```diff
@@ -1,4 +1,5 @@
-﻿using System.Diagnostics;
+﻿using Ryujinx.Profiler;
+using System.Diagnostics;
```
Comment (gdkchan) [inline Ryujinx.HLE/PerformanceStatistics.cs]: Is this really needed?

File: Ryujinx.Profiler/DumpProfile.cs [hunk 35]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace Ryujinx.Profiler
+{
+    public static class DumpProfile
+    {
+        public static void ToFile(string path, InternalProfile profile)
+        {
+            String fileData = "Category,Session Group,Session Item,Count,Average(ms),Total(ms)\r\n";
+
+            foreach (var time in profile.Timers.OrderBy(key => key.Key.Tag))
```
Comment (gdkchan) [inline Ryujinx.Profiler/DumpProfile.cs]: I don't like the user of var here as its not easy to see the type of `time` without looking at the definition on `InternalProfile`.

File: Ryujinx.Profiler/InternalProfile.cs [hunk 36]
```diff
@@ -0,0 +1,224 @@
...
+                Config  = config,
+                IsBegin = false,
+                Time    = PerformanceCounter.ElapsedTicks,
+            });
+        }
+
+        public string GetSession()
+        {
+            // Can be called from multiple threads so locked to ensure no duplicate sessions are generated
+            lock (_sessionLock)
```
Comment (gdkchan) [inline Ryujinx.Profiler/InternalProfile.cs]: You don't need a lock for that, you can use `Interlocked.Increment` to increment the value atomically.

File: Ryujinx.Profiler/NPadDebug.cs [hunk 37]
```diff
@@ -0,0 +1,31 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Text;
+using OpenTK.Input;
+
+namespace Ryujinx.Profiler
+{
+    public struct NpadDebugButtons
```
Comment (gdkchan) [inline Ryujinx.Profiler/NPadDebug.cs]: Why is it called `Npad`? I know that on the HID service, Npad is usually used to refer to the different controller types, but this seems completely unrelated, so it's not clear to me why it is called `NpadDebug`.

File: Ryujinx.Profiler/Profile.cs [hunk 38]
```diff
@@ -0,0 +1,143 @@
...
+{
+    public static class Profile
+    {
+        public static float     UpdateRate    => _settings.UpdateRate;
+        public static long      HistoryLength => _settings.History;
+        public static NpadDebug Controls      => _settings.Controls;
+
+        private static InternalProfile  _profileInstance;
+        private static ProfilerSettings _settings;
+
```
Comment (gdkchan) [inline Ryujinx.Profiler/Profile.cs]: nit: Unneeded empty new line.

File: Ryujinx.Profiler/Profile.cs [hunk 39]
```diff
@@ -0,0 +1,143 @@
...
+                History         = (long)(config.History * PerformanceCounter.TicksPerSecond),
+                MaxLevel        = config.MaxLevel,
+                Controls        = config.Controls,
+                MaxFlags        = config.MaxFlags,
+            };
+        }
+
+        public static bool ProfilingEnabled()
+        {
+            #if USE_PROFILING
```
Comment (gdkchan) [inline Ryujinx.Profiler/Profile.cs]: Usually those #if #else #endif etc directives are not idented. As it is now, it's imo a bit hard to see at a first glance what is supposed to run on each condition.

File: Ryujinx.Profiler/ProfileConfig.cs [hunk 40]
```diff
@@ -0,0 +1,121 @@
...
+                        _cachedSession = "";
+                    }
+                }
+
+                return _cachedSession;
+            }
+        }
+
+        /// <summary>
+        /// This equals overload is vital
```
Comment (gdkchan) [inline Ryujinx.Profiler/ProfileConfig.cs]: I don't think that `This equals overload is vital` is necessary on the summary.

File: Ryujinx.Profiler/ProfileConfig.cs [hunk 41]
```diff
@@ -0,0 +1,121 @@
...
+        /// The default comparison is far too slow for the number of comparisons needed because it doesn't know what's important to compare
+        /// </summary>
+        /// <param name="obj">Object to compare to</param>
+        /// <returns></returns>
+        public override bool Equals(object obj)
```
Comment (gdkchan) [inline Ryujinx.Profiler/ProfileConfig.cs]: May be worth to implement `IEquatable<ProfileConfig>` and add a equals overload accepting `ProfileConfig` to avoid boxing in some cases.

File: Ryujinx.Profiler/ProfilerConfig.jsonc [hunk 42]
```diff
@@ -0,0 +1,28 @@
+﻿{
+  // Enable profiling (Only available on a profiling enabled build)
```
Comment (gdkchan) [inline Ryujinx.Profiler/ProfilerConfig.jsonc]: nit: `profiling enabled build` -> `profiling enabled builds`.

File: Ryujinx.Profiler/Settings.cs [hunk 43]
```diff
@@ -0,0 +1,23 @@
...
+        // Default settings for profiler
+        public bool   Enabled         { get; set; } = false;
+        public bool   FileDumpEnabled { get; set; } = false;
+        public string DumpLocation    { get; set; } = "";
+        public float  UpdateRate      { get; set; } = 0.1f;
+        public int    MaxLevel        { get; set; } = 0;
+        public int    MaxFlags        { get; set; } = 1000;
+
+        // 19531225 = 5 seconds in ticks
+        public long   History { get; set; } = 19531225;
```
Comment (gdkchan) [inline Ryujinx.Profiler/Settings.cs]: I beleive the extra spacing between long and History is not necessary. Also I believe the counter frequency *may* vary from system to system. So while "19531225" is equivalent to 5 seconds on your system, it may not be in other systems.

File: Ryujinx.Profiler/TimingFlag.cs [hunk 44]
```diff
@@ -0,0 +1,22 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Drawing;
+using System.Text;
+
+namespace Ryujinx.Profiler
+{
+    public enum TimingFlagType
```
Comment (gdkchan) [inline Ryujinx.Profiler/TimingFlag.cs]: Why it this enum explicitly numbered?

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 45]
```diff
@@ -0,0 +1,772 @@
...
+            Pause             = 6,
+            ChangeDisplay     = 7,
+
+            // Don't automatically draw after here
+            ToggleFlags       = 8,
+            Step              = 9,
+
+            // Update this when new buttons are added
+            Count             = 10,
+            Autodraw          = 8,
```
Comment (gdkchan) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: Was this a mistake? Value 8 is already used by `ToggleFlags`. If not, I think would be nice to add a comment here.

File: Ryujinx.Profiler/UI/ProfileWindow.cs [hunk 46]
```diff
@@ -0,0 +1,772 @@
...
+            _visibleChanged = true;
+        }
+
+        private void SetSort(IComparer<KeyValuePair<ProfileConfig, TimingInfo>> filter)
+        {
+            _sortAction = filter;
+            _profileUpdated = true;
+        }
+
+        #region OnLoad
```
Comment (gdkchan) [inline Ryujinx.Profiler/UI/ProfileWindow.cs]: iirc #region is usually also not idented.

File: Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs [hunk 47]
```diff
@@ -0,0 +1,257 @@
...
+        private int CharacterTextureSheet;
+        private CharacterInfo[] characters;
+
+        public Color fontColor { get; set; } = Color.Black;
+
+        private string GetFontPath()
+        {
+            string fontFolder = System.Environment.GetFolderPath(Environment.SpecialFolder.Fonts);
+
+            // Only uses Arial, add more font here if wanted
```
Comment (gdkchan) [inline Ryujinx.Profiler/UI/SharpFontHelpers/FontService.cs]: `more font` -> `more fonts`.

File: Ryujinx/Program.cs [hunk 48]
```diff
@@ -5,6 +5,7 @@
 using Ryujinx.HLE;
 using System;
 using System.IO;
+using Ryujinx.Profiler;
```
Comment (gdkchan) [inline Ryujinx/Program.cs]: nit: Sorting.

File: Ryujinx.Profiler/NPadDebug.cs (hunk 37 above)
Comment (BaronKiko) [reply Ryujinx.Profiler/NPadDebug.cs]: Fair point. Refactored out

File: Ryujinx.Profiler/Settings.cs (hunk 43 above)
Comment (BaronKiko) [reply Ryujinx.Profiler/Settings.cs]: I fixed the spacing
The time equals that number on any pc with a high precision clock which I have been led to believe is most computers.
It shouldn't really matter anyway because it sets that value on boot with the time specified in config. I have updated the comment to clarify this.

File: Ryujinx.Profiler/UI/ProfileWindow.cs (hunk 45 above)
Comment (BaronKiko) [reply Ryujinx.Profiler/UI/ProfileWindow.cs]: These 2 values are used as indexes in the list. This is also why they are explicitly numbered.
Up to auto draw it will automatically do the rendering of the buttons with a single loop. After that it draws at various points in the code. This is so it can control visibility of the buttons and update positions.

File: Ryujinx.Profiler/InternalProfile.cs (hunk 36 above)
Comment (BaronKiko) [reply Ryujinx.Profiler/InternalProfile.cs]: Ah cool, I didn't know about that.

File: Ryujinx.HLE/PerformanceStatistics.cs (hunk 34 above)
Comment (BaronKiko) [reply Ryujinx.HLE/PerformanceStatistics.cs]: It's needed for the stopwatch class

File: Ryujinx.Profiler/Profile.cs (hunk 39 above)
Comment (BaronKiko) [reply Ryujinx.Profiler/Profile.cs]: I can unindent it. I usually keep them indented just because I think it looks nicer

**Comment by BaronKiko** [comment]
Fixed all review comments

File: Ryujinx.Profiler/InternalProfile.cs [hunk 49]
```diff
@@ -0,0 +1,221 @@
...
+            public long Time;
+            public bool IsBegin;
+        }
+
+        internal Dictionary<ProfileConfig, TimingInfo> Timers { get; set; }
+
+        private readonly object _timerQueueClearLock = new object();
+        private ConcurrentQueue<TimerQueueValue> _timerQueue;
+
+        private readonly object _sessionLock = new object();
```
Comment (gdkchan) [inline Ryujinx.Profiler/InternalProfile.cs]: I guess this is now unused?

Comment (BaronKiko) [reply Ryujinx.Profiler/InternalProfile.cs]: Quite right. Removed

**Comment by BaronKiko** [comment]
I have tested it on my system. Would be nice to see performance on a weaker system but I really don't see how I could optimize it further so I'm not sure what I could do to help that.
It's all stripped out on normal builds so it shouldn't effect normal users.

I missed these comments
> I see that you wrote a whole GUI using OpenTK for the profiler, which is nice but we're going to need rewriting this portion of the code once we have a GUI on the emulator

It will be tough to keep up for most ui's but we can come to that when a ui happens

>  it is using legacy OpenGL

I probably should have used modern gl but at the time I chose to go for maximum compatibility. Modern ogl wasn't needed and traditional opengl allows some optimizations that would be harder to make, perhaps impossible.
